### ISSUE TITLE
Native 

### ISSUE THREAD

**Issue by mvysny** [mvysny/vaadin-boot#10]
Native is cool these days. Add support for building with Graalvm; there's a Gradle plugin to build for native. Jetty and class auto discovery needs to be figured out.

**Comment by mvysny** [comment]
I've installed GraalVM and went through the tutorial at https://graalvm.github.io/native-build-tools/latest/gradle-plugin-quickstart.html - everything worked well.

Tried the same steps (agent+metadataCopy+nativeCompile) on https://github.com/mvysny/vaadin-boot-example-gradle, but unfortunately I'm getting

```
java.lang.NullPointerException
        at com.vaadin.flow.server.VaadinServlet.serveStaticOrWebJarRequest(VaadinServlet.java:431)
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:386)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1665)
        at org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:170)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:527)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:578)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
```
which means that `VaadinServlet.staticFileHandler` is null, which means that `VaadinServlet.init()` wasn't called. The startup time is instantaneous though - 17ms - which looks really promising.

I've tried to add the `[Lorg.eclipse.jetty.servlet.ServletMapping;` mapping as per https://javalin.io/2018/09/27/javalin-graalvm-example.html , didn't help. The tutorial looks old so it's probably of no value now. Maybe an updated tutorial of javalin+graalvm could also help.

I'm compiling Vaadin dev mode to native atm which is kind of dumb. I'll try the same thing but with prod mode.

**Comment by mvysny** [comment]
Unfortunately the production build fails with the same exception.

Further things to check out:

* https://github.com/marcushellberg/flow-aot-demo
* https://docs.spring.io/spring-boot/docs/current/reference/html/native-image.html
* https://github.com/eclipse/jetty.project/issues/3095 - unfortunately it's stale

**Comment by mvysny** [comment]
Debugged VaadinServlet in native a bit (by introducing my own MyServlet which extends VaadinServlet, and adding a bunch of stdout statements), found interesting info. `VaadinServlet.init()` actually does get called, however it soon bails out and does nothing, because `Lookup` is `null`. According to https://mvysny.github.io/vaadin-lookup-vs-instantiator/ that means that `LookupServletContainerInitializer` is not called, which means that Jetty classpath autodiscovery isn't working properly.

**Comment by mvysny** [comment]
Could the Quick start be the solution? https://github.com/mvysny/vaadin-boot/issues/11

**Comment by mvysny** [comment]
Yup, QuickStart should help. Blocked by #11

**Comment by mvysny** [comment]
When running the native binary, QuickStart fails with

```
2023-03-16 13:32:47.747 [main] WARN org.eclipse.jetty.webapp.WebAppContext - Failed startup of context o.e.j.w.WebAppContext@1da6646c{/,resource:/webapp,STOPPED}{resource:/webapp}
java.lang.IllegalStateException: Bad Quickstart location
	at org.eclipse.jetty.quickstart.QuickStartConfiguration.preConfigure(QuickStartConfiguration.java:95)
	at org.eclipse.jetty.webapp.Configurations.preConfigure(Configurations.java:496)
	at org.eclipse.jetty.webapp.WebAppContext.preConfigure(WebAppContext.java:510)
```

**Comment by mvysny** [comment]
The reason for that is that `context.getBaseResource()` resolves to `resource:/webapp` of type `URLResource`; both `URLResource.isDirectory()` and `URLResource.exists()` return false. I've patched `QuickStartConfiguration:94` and commented out the check. I've also passed in the path to `quickstart-web.xml` as `ctx.setAttribute(QuickStartConfiguration.QUICKSTART_WEB_XML, new PathResource(new File("src/main/resources/webapp/WEB-INF/quickstart-web.xml")));`, and now the native mode works.

I'll polish the native mode, document, file a feature request at Jetty.

**Comment by mvysny** [comment]
Blocked by https://github.com/eclipse/jetty.project/issues/9514

**Comment by mvysny** [comment]
Closing as `wontfix`. GraalVM native-image is out of scope for vaadin-boot, and the cost/benefit is upside-down.

## Why the benefit is marginal *for this workload*

Native-image's selling points are fast cold start and low RSS — which pay off for **serverless / scale-to-zero / CLI** workloads. A Vaadin Flow app is the opposite: a **long-running, stateful server** that holds `VaadinSession`s in memory and amortizes JVM warmup over hours/days. You pay JIT warmup once; native buys you back a few seconds on a process that lives for weeks. The one metric where native could help (idle memory footprint) isn't where Vaadin apps hurt — session state and the component tree dominate, and that's identical native or not. So the headline win doesn't apply to the thing vaadin-boot boots.

## Why the cost is large and *recurring* (not one-time)

1. **Flow is reflection- and classpath-scan-heavy at runtime** — route discovery, component/`@Route`/`@WebServlet`/`@WebListener` auto-discovery, DI. Native image closes the world at build time, so every one of those needs reachability metadata (`reflect-config`, `resource-config`, `serialization-config`, proxy config). Vaadin's *own* native story is delivered **through Spring Boot's** metadata plumbing — and vaadin-boot's entire reason to exist is "no Spring." We'd be reimplementing that metadata layer by hand.

2. **Jetty embedded annotation scanning** (`ServletDeployer`, WebSocket upgrade) is exactly the "class auto-discovery" problem, and it's the part that fights native hardest. Same for Tomcat. That's two container backends to maintain metadata for, not one.

3. **It breaks on every Vaadin bump.** This is the killer, and it's why this was labeled `blocked by upstream`. Reachability metadata tracks internal Flow classes that change between minor versions; a routine Vaadin bump silently breaks the native build **at runtime, not compile time**. That means a native CI job on the release gate and debugging `ClassNotFoundException`s in a stripped binary — a permanent tax on a library whose whole appeal is that it's tiny and boring.

4. **Native doesn't even cover dev mode.** You'd only ever native the production build; `npm`/Node frontend compilation still happens the normal way. So the supported surface would be "production-only, per-container, per-Vaadin-version" — maximum fragility, minimum coverage.

## If you genuinely need native

Use **Vaadin + Spring Boot native**, which is the officially supported path. vaadin-boot's niche is "the simplest possible non-Spring boot," and that niche is orthogonal to native image.
